src/MongoDB/Session.c

```diff
@@ -228,6 +230,154 @@ static PHP_METHOD(Session, getOperationTime)
 	php_phongo_new_timestamp_from_increment_and_timestamp(return_value, increment, timestamp TSRMLS_CC);
 } /* }}} */
 
+/* Creates a opts structure from an array optionally containing an RP, RC,
+ * and/or WC object. Returns NULL if no options were found, or there was an
+ * invalid option. If there was an invalid option or structure, an exception
+ * will be thrown too. */
+mongoc_transaction_opt_t* php_mongodb_session_parse_transaction_options(zval* options TSRMLS_DC)
+{
+	mongoc_transaction_opt_t* opts = NULL;
+
+	if (php_array_existsc(options, "readConcern")) {
+		zval* read_concern = php_array_fetchc(options, "readConcern");
+
+		if (Z_TYPE_P(read_concern) != IS_OBJECT || !instanceof_function(Z_OBJCE_P(read_concern), php_phongo_readconcern_ce TSRMLS_CC)) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected \"readConcern\" option to be %s, %s given", ZSTR_VAL(php_phongo_readconcern_ce->name), PHONGO_ZVAL_CLASS_OR_TYPE_NAME_P(read_concern));
+			return false;
+		}
+
+		if (!opts) {
+			opts = mongoc_transaction_opts_new();
+		}
+
+		mongoc_transaction_opts_set_read_concern(opts, phongo_read_concern_from_zval(read_concern TSRMLS_CC));
+	}
+
+	if (php_array_existsc(options, "readPreference")) {
+		zval* read_preference = php_array_fetchc(options, "readPreference");
+
+		if (Z_TYPE_P(read_preference) != IS_OBJECT || !instanceof_function(Z_OBJCE_P(read_preference), php_phongo_readpreference_ce TSRMLS_CC)) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected \"readPreference\" option to be %s, %s given", ZSTR_VAL(php_phongo_readpreference_ce->name), PHONGO_ZVAL_CLASS_OR_TYPE_NAME_P(read_preference));
+			return false;
+		}
+
+		if (!opts) {
+			opts = mongoc_transaction_opts_new();
+		}
+
+		mongoc_transaction_opts_set_read_prefs(opts, phongo_read_preference_from_zval(read_preference TSRMLS_CC));
+	}
+
+	if (php_array_existsc(options, "writeConcern")) {
+		zval* write_concern = php_array_fetchc(options, "writeConcern");
+
+		if (Z_TYPE_P(write_concern) != IS_OBJECT || !instanceof_function(Z_OBJCE_P(write_concern), php_phongo_writeconcern_ce TSRMLS_CC)) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected \"writeConcern\" option to be %s, %s given", ZSTR_VAL(php_phongo_writeconcern_ce->name), PHONGO_ZVAL_CLASS_OR_TYPE_NAME_P(write_concern));
+			return false;
```

Comment:
Technically, it may be possible for this to leak if a valid ReadConcern or ReadPreference is added to `mongoc_transaction_opt_t* opts` but the WriteConcern fails. Likewise for a valid ReadConcern and invalid ReadPreference. Both of these cases should be easy to test and demonstrate a leak. I imagine the current error tests didn't expose this because the test cases each contain one option at a time.

Should just require adding the following here and in ReadPreference:

```c
if (opts) {
	mongoc_transaction_opts_destroy(opts);
}
```

Reply:
Done